app/setup_handlers.go

```diff
@@ -52,26 +57,72 @@ func SetupHandlers(app *App) {
 		}
 	}
 	baseAppLegacySS := app.ParamsKeeper.Subspace(baseapp.Paramspace).WithKeyTable(paramstypes.ConsensusParamsKeyTable())
-	app.UpgradeKeeper.SetUpgradeHandler(constant.Version, func(ctx sdk.Context, plan types.Plan, vm module.VersionMap) (module.VersionMap, error) {
-		app.Logger().Info("Running upgrade handler for " + constant.Version)
-		// Migrate Tendermint consensus parameters from x/params module to a dedicated x/consensus module.
-		baseapp.MigrateParams(ctx, baseAppLegacySS, &app.ConsensusParamsKeeper)
-		// Updated version map to the latest consensus versions from each module
-		for m, mb := range app.mm.Modules {
-			if module, ok := mb.(module.HasConsensusVersion); ok {
-				vm[m] = module.ConsensusVersion()
-			}
-		}
+	needsForcedMigration := []string{
+		authtypes.ModuleName,
+		banktypes.ModuleName,
+		stakingtypes.ModuleName,
+		distrtypes.ModuleName,
+		slashingtypes.ModuleName,
+		govtypes.ModuleName,
+		crisistypes.ModuleName,
+		emissionstypes.ModuleName,
+	}
+	allUpgrades := upgradeTracker{
+		upgrades: []upgradeTrackerItem{
+			{
+				index: 1714664193,
+				storeUpgrade: &storetypes.StoreUpgrades{
+					Added: []string{consensustypes.ModuleName, crisistypes.ModuleName},
+				},
+				upgradeHandler: func(ctx sdk.Context, vm module.VersionMap) (module.VersionMap, error) {
+					// Migrate Tendermint consensus parameters from x/params module to a dedicated x/consensus module
+					// https://docs.cosmos.network/main/build/migrations/upgrading#xconsensus
+					baseapp.MigrateParams(ctx, baseAppLegacySS, &app.ConsensusParamsKeeper)
+
+					// empty version map happens when upgrading from old versions which did not correctly call
+					// app.UpgradeKeeper.SetModuleVersionMap(ctx, app.mm.GetVersionMap()) in InitChainer.
+					// we must populate the version map if we detect this scenario
+					//
+					// this will only happen on the first upgrade. mainnet and testnet will not require this condition.
+					if len(vm) == 0 {
+						for m, mb := range app.mm.Modules {
+							if module, ok := mb.(module.HasConsensusVersion); ok {
+								if slices.Contains(needsForcedMigration, m) {
+									vm[m] = module.ConsensusVersion() - 1
+								} else {
+									vm[m] = module.ConsensusVersion()
+								}
+							}
+						}
+					}
+					return vm, nil
+				},
+			},
+			{
+				index: 1715624665,
+				storeUpgrade: &storetypes.StoreUpgrades{
+					Added: []string{capabilitytypes.ModuleName, ibcexported.ModuleName, ibctransfertypes.ModuleName},
+				},
+			},
+		},
+	}
+
+	_, useDevelopTracker := os.LookupEnv("ZETACORED_USE_DEVELOP_UPGRADE_TRACKER")
+	upgradeHandlerFns, storeUpgrades, err := allUpgrades.getUpgrades(useDevelopTracker)
+	if err != nil {
+		panic(err)
```

Comment:
Do we need this panic here? Can we return instead ?

Reply:
Sure I could refactor `func SetupHandlers(app *App)` to `func SetupHandlers(app *App) error`. But we'd still just panic in `app.New()` anyway right?